linker/shared/src/main/scala/org/scalajs/linker/frontend/LambdaSynthesizer.scala

```diff
@@ -0,0 +1,177 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package org.scalajs.linker.frontend
+
+import org.scalajs.ir.{ClassKind, Position, SHA1, UTF8String, Version}
+import org.scalajs.ir.Names._
+import org.scalajs.ir.OriginalName.NoOriginalName
+import org.scalajs.ir.Trees._
+import org.scalajs.ir.Types._
+import org.scalajs.ir.WellKnownNames._
+
+import org.scalajs.linker.analyzer.Infos._
+
+private[linker] object LambdaSynthesizer {
+  /* Everything we create has a constant version because the names are derived
+   * from the descriptors themselves.
+   */
+  val constantVersion = Version.fromByte(0)
+
+  private val ClosureTypeRefName = LabelName("c")
+
+  /** Deterministically makes a class name for the lambda class given its descriptor.
+   *
+   *  This computation is mildly expensive. Callers should cache it if possible.
+   */
+  def makeClassName(descriptor: NewLambda.Descriptor): ClassName = {
+    // Choose a base class name that will "makes sense" for debugging purposes
+    val baseClassName = {
+      if (descriptor.superClass == ObjectClass && descriptor.interfaces.nonEmpty)
+        descriptor.interfaces.head
+      else
+        descriptor.superClass
+    }
+
+    val digestBuilder = new SHA1.DigestBuilder()
+    digestBuilder.updateUTF8String(descriptor.superClass.encoded)
+    for (intf <- descriptor.interfaces)
+      digestBuilder.updateUTF8String(intf.encoded)
+
+    // FIXME This is not efficient
+    digestBuilder.updateUTF8String(UTF8String(descriptor.methodName.nameString))
+
+    // No need the hash the paramTypes and resultType because they derive from the method name
+
+    val digest = digestBuilder.finalizeDigest()
+
+    /* The "$$Lambda" segment is meant to match the way LambdaMetaFactory
+     * names generated classes. This is mostly for test compatibility
+     * purposes (like partest's that test the class name to tell whether a
+     * lambda was indeed encoded as an LMF).
+     */
+    val suffixBuilder = new java.lang.StringBuilder(".$$Lambda$")
+    for (b <- digest) {
+      val i = b & 0xff
+      suffixBuilder.append(Character.forDigit(i >> 4, 16)).append(Character.forDigit(i & 0x0f, 16))
+    }
+
+    ClassName(baseClassName.encoded ++ UTF8String(suffixBuilder.toString()))
+  }
+
+  /** Computes the constructor name for the lambda class of a descriptor. */
+  def makeConstructorName(descriptor: NewLambda.Descriptor): MethodName = {
+    val closureTypeNonNull =
+      ClosureType(descriptor.paramTypes, descriptor.resultType, nullable = false)
+    MethodName.constructor(TransientTypeRef(ClosureTypeRefName)(closureTypeNonNull) :: Nil)
```

Comment:
Comment corresponding to the comment in `Types`: This violates the constraint that same name implies same type for transient type refs.

Reply:
Following up on the comment above: this does not violate the constraint, because the constraint is not global.